Update ReadME + Examples (#2)
RenaudLN/dash_socketio: README.md

# Dash Socketio

This package integration of the [Flask SocketIO](https://flask-socketio.readthedocs.io/en/latest/) into Dash.
Flask SocketIO provides the server-side capabilities of pushing data to clients, while Dash SocketIO enables the Dashboard to recieve and process the messages.

## Install

```shell
pip install dash_socketio
```
## Use Cases:
There are several potential use cases which Dash SocketIO enables:
1. Send update messages while a callback is processing, e.g. streaming responses or downloading data
2. Sending messages to multiple users at once and across clients, e.g. chat applications

> 3. Align data or filters for the same user across clients.

Author: In my current project, we have two pages with the same dropdown. We got the request to implement a feature where the user opened the two different pages (either two tabs same brwoser or even two different browsers). If the selects a value in one dropdown, for example "NewYork", he wants to update it simultaneously on the other tab/browser and change the content of the page accordingly. 

It's a way to keep the values of the dropdown in sync. Clients would mean, tabs, browsers, or windows. As long as it's the same user he gets "NewYotk"

> 3. Align data or filters for the same user across clients.

Author: https://user-images.githubusercontent.com/913165/82079294-3cfb5d80-96a8-11ea-99be-35473d28a5f2.gif

Similar to this preview.

)
def display_status(n_clicks, socket_id):
    if not n_clicks or not socket_id:
        return no_update, []
    notify(socket_id, "Sending the question to LoremLM...")
    time.sleep(1)
    notify(socket_id, "Streaming answer...")

    for i, word in enumerate(paragraph.replace("\n", " ").split(" ")):
        emit("stream", " " * bool(i) + word, namespace="/", to=socket_id)
        time.sleep(0.05)

    notify(socket_id, "Done!", color="green")

    return paragraph, []

clientside_callback(
    """connected => !connected""",
    Output("btn", "disabled"),
    Input("socketio", "connected"),
)

clientside_callback(
    """(notification) => {
        if (!notification) return dash_clientside.no_update
        return notification
    }""",
    Output("notification_wrapper", "children", allow_duplicate=True),
    Input("socketio", "data-notification"),
    prevent_initial_call=True,
)

clientside_callback(
    """(word, text) => text + word""",
    Output("results", "children", allow_duplicate=True),
    Input("socketio", "data-stream"),
    State("results", "children"),
    prevent_initial_call=True,
)

if __name__ == '__main__':
    app.run_server(debug=True, dev_tools_ui=False)

> ```

Author: Sure, will you add the gif later? I will delete the code

## SocketIO Reference

```
- id (string; optional):
    Unique ID to identify this component in Dash callbacks.

- connected (boolean; optional):
    Whether the client is connected to the websocket - READONLY.

- eventNames (list of strings; optional):
    Name of Socket.IO events to listen to.

- socketId (string; optional):
    The socket ID.

- url (string; optional):
    The socket.io server URL."""

```

> The events can be accessed in a clientsside callback via:
> ` Input("socketio", "data-eventName")`
> Note that the event needs top trigger a client-side callback.

Reviewer: It doesn't need to be clientside callbacks. In a lot of cases it will make sense for it to be with clientside callbacks though.
```suggestion
The events can be accessed in a Dash callbacks (client- or server-side) via:
` Input(<id>, "data-<eventName>")`
```

Author: Sure. Shall we add something to show what could be done if the server access the event? So far I used your emit a server side action and update on the client side. I did not have the use atm, where the client, or DashSocketIO, did sent something to the server.
